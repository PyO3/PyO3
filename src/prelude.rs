//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
