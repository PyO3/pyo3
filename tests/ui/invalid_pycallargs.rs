use pyo3::prelude::*;

fn main() {
    Python::with_gil(|py| {
        let any = py.None().into_bound(py);
        any.call1("foo");
    })
}
