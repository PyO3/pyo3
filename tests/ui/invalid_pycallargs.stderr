error[E0277]: `&str` cannot used as a Python `call` argument
 --> tests/ui/invalid_pycallargs.rs:6:19
  |
6 |         any.call1("foo");
  |             ----- ^^^^^ the trait `PyCallArgs<'_>` is not implemented for `&str`
  |             |
  |             required by a bound introduced by this call
  |
  = note: `PyCallArgs` is implemented for Rust tuples, `Bound<'py, PyTuple>` and `Py<PyTuple>`
  = note: if your type is convertable to `PyTuple` via `IntoPyObject`, call `<arg>.into_pyobject(py)` manually
  = note: if you meant to pass the type as a single argument, wrap it in a 1-tuple, `(<arg>,)`
  = help: the following other types implement trait `PyCallArgs<'py>`:
            &'a (T0, T1)
            &'a (T0, T1, T2)
            &'a (T0, T1, T2, T3)
            &'a (T0, T1, T2, T3, T4)
            &'a (T0, T1, T2, T3, T4, T5)
            &'a (T0, T1, T2, T3, T4, T5, T6)
            &'a (T0, T1, T2, T3, T4, T5, T6, T7)
            &'a (T0, T1, T2, T3, T4, T5, T6, T7, T8)
          and $N others
note: required by a bound in `call1`
 --> src/types/any.rs
  |
  |     fn call1<A>(&self, args: A) -> PyResult<Bound<'py, PyAny>>
  |        ----- required by a bound in this associated function
  |     where
  |         A: PyCallArgs<'py>;
  |            ^^^^^^^^^^^^^^^ required by this bound in `PyAnyMethods::call1`
help: use a unary tuple instead
  |
6 |         any.call1(("foo",));
  |                   +     ++
